File: LeetCode/src/greedy/IncreasingTripletSubsequence.java
package greedy;

public class IncreasingTripletSubsequence {

    public boolean increasingTriplet(int[] nums) {
        if (nums.length < 3) {
            return false;
        }

        int first = Integer.MAX_VALUE;
        int second = Integer.MAX_VALUE;

        for (int num : nums) {
            if (num <= first) {
                first = num;
            } else if (num <= second) {
                second = num;
            } else {
                return true;
            }
        }


        return false;
    }

    public static void main(String[] args) {
        IncreasingTripletSubsequence solution = new IncreasingTripletSubsequence();
        System.out.println(solution.increasingTriplet(new int[]{1, 2, 3, 4, 5})); // true
        System.out.println(solution.increasingTriplet(new int[]{5, 4, 3, 2, 1})); // false
        System.out.println(solution.increasingTriplet(new int[]{2, 1, 5, 0, 4, 6})); // true
    }
}

File: LeetCode_Python/array/IncreasingTripletSubsequence.py
class IncreasingTripletSubsequence:
    def increasingTriplet(self, nums: list[int]) -> bool:
        if len(nums) < 3:
            return False

        first = second = float("inf")

        for n in nums:
            if n <= first:
                first = n
            elif n <= second:
                second = n
            else:
                return True

        return False


# Time Complexity: O(n), where n is the length of the input list
# Space Complexity: O(1), since we are using only a constant amount of space

if __name__ == "__main__":
    solution = IncreasingTripletSubsequence()

    print("Expected Output: True")
    print("Actual Output:", solution.increasingTriplet([1, 2, 3, 4, 5]))

    print("Expected Output: False")
    print("Actual Output:", solution.increasingTriplet([5, 4, 3, 2, 1]))

    print("Expected Output: True")
    print("Actual Output:", solution.increasingTriplet([2, 1, 5, 0, 4, 6]))
